openebs/openebs-provisioner.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"os"
+
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	mApiv1 "github.com/kubernetes-incubator/external-storage/openebs/pkg/v1"
+	mayav1 "github.com/kubernetes-incubator/external-storage/openebs/types/v1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	provisionerName = "openebs.io/provisioner-iscsi"
+)
+
+type openEBSProvisioner struct {
+	// Maya-API Server URI running in the cluster
+	mapiURI string
+
+	// Identity of this openEBSProvisioner, set to node's name. Used to identify
+	// "this" provisioner's PVs.
+	identity string
+}
+
+// NewOpenEBSProvisioner creates a new openebs provisioner
+func NewOpenEBSProvisioner(client kubernetes.Interface) controller.Provisioner {
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		glog.Fatal("env variable NODE_NAME must be set so that this provisioner can identify itself")
+	}
+	var openebsObj mApiv1.OpenEBSVolume
+
+	//Get maya-apiserver IP address from cluster
+	addr, err := openebsObj.GetMayaClusterIP(client)
+
+	if err != nil {
+		glog.Fatalf("Error getting maya-api-server IP Address: %v", err)
+		return nil
+	}
+	mayaServiceURI := "http://" + addr + ":5656"
+
+	//Set maya-apiserver IP address along with default port
+	os.Setenv("MAPI_ADDR", mayaServiceURI)
+
+	return &openEBSProvisioner{
+		mapiURI:  mayaServiceURI,
+		identity: nodeName,
+	}
+}
+
+var _ controller.Provisioner = &openEBSProvisioner{}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *openEBSProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+
+	//Issue a request to Maya API Server to create a volume
+	var volume mayav1.Volume
+	var openebsVol mApiv1.OpenEBSVolume
+
+	volSize := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+
+	_, err := openebsVol.CreateVsm(options.PVName, volSize.String())
+	if err != nil {
+		glog.Fatalf("Error creating volume: %v", err)
+		return nil, err
+	}
+
+	err = openebsVol.ListVsm(options.PVName, &volume)
```

Comment:
Ideally `createVSM` should return an error/success. The logic should be there in `createVSM` to timeout and error to caller if it didnt happen within the allowed slot. In a different thought, you could implement `volumeinfo` kind of call for a specific volume detail and it should help instead of complete listing of volume..etc.